File: src/main/java/rip/sunrise/injectapi/global/ProxyDynamicFactory.java
package rip.sunrise.injectapi.global;

import java.lang.invoke.*;

public class ProxyDynamicFactory {
    /**
     * The InjectAPI ClassLoader.
     * Instantly set after class definition.
     */
    public static ClassLoader classLoader;

    /**
     * Used from within the internal hook code to make sure we are not going into recursion.
     * Instantly set after class definition, reset on transform.
     */
    @SuppressWarnings("unused")
    public static ThreadLocal<boolean[]> runningHooks = ThreadLocal.withInitial(() -> new boolean[0]);

    /**
     * Called from hooks. This is the bridge between any ClassLoader and the InjectAPI ClassLoader.
     * Because this is only called once. It's okay to use reflection here.
     */
    public static CallSite bootstrap(MethodHandles.Lookup caller, String name, MethodType type, int hookId) {
        return new ConstantCallSite(getHandle(hookId).asType(type));
    }

    @SuppressWarnings("ReplaceOnLiteralHasNoEffect")
    public static MethodHandle getHandle(int hookId) {
        try {
            Class<?> managerClass = classLoader.loadClass("@HOOK_MANAGER@".replace("/", "."));
            Class<?> hookClass = classLoader.loadClass("@HOOK@".replace("/", "."));

            Object instance = managerClass.getDeclaredField("INSTANCE").get(null);

            Object hook = managerClass.getDeclaredMethod("getCachedHook", int.class).invoke(instance, hookId);

            return (MethodHandle) hookClass.getDeclaredMethod("getHandle").invoke(hook);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/rip/sunrise/injectapi/callsite/ProxyDynamicFactory.kt
package rip.sunrise.injectapi.callsite

import rip.sunrise.injectapi.hooks.inject.Context
import java.lang.invoke.*

/**
 * Class containing a bootstrap method for all the hooks.
 *
 * NOTE: Loaded on ALL hooked ClassLoaders.
 */
object ProxyDynamicFactory {
    /**
     * This method is called as a bootstrap for invokedynamic in hooked code. The returned [CallSite] points to the hooks [MethodHandle].
     */
    @JvmStatic
    fun bootstrap(caller: MethodHandles.Lookup, name: String, type: MethodType, hookId: Int): CallSite {
        try {
            println("Got invokedynamic call from $caller for hookId $hookId")

            val cl = ClassLoader.getSystemClassLoader()
                .loadClass("rip.sunrise.injectapi.callsite.DataTransport")
                .getDeclaredMethod("getClassLoader")
                .invoke(null) as ClassLoader

            val clazz = cl.loadClass("rip.sunrise.injectapi.managers.HookManager")
            val hookClass = cl.loadClass("rip.sunrise.injectapi.hooks.Hook")

            // Note: DO NOT CAST. This is most likely not the correct classloader. current != cl
            val instance = clazz.getDeclaredField("INSTANCE").get(null)

            val hook = clazz.getDeclaredMethod("getHook", Int::class.java).invoke(instance, hookId)
            val handle = hookClass.getDeclaredMethod("getHandle").invoke(hook) as MethodHandle

            // TODO: Kinda forced, it won't fail. Only happens because types are lost when calling unreflect.
            return ConstantCallSite(handle.asType(type))
        } catch (e: Exception) {
            e.printStackTrace()
            error(e)
        }
    }
}